src/cpp/rtps/messages/submessages/DataMsg.hpp

```diff
@@ -147,11 +147,18 @@ struct DataMsgUtils
                 msg, change->write_params.original_writer_info());
         }
 
-        if (WITH_KEY == topicKind && (!change->writerGUID.is_builtin() || expectsInlineQos || ALIVE != change->kind))
+        if (WITH_KEY == topicKind && change->instanceHandle.isDefined() &&
+                (!change->writerGUID.is_builtin() || expectsInlineQos || ALIVE != change->kind))
         {
+            /**
+             * If instanceHandle is not defined, this means the key hash is not populated. It makes no sense
+             * to serialize a parameter with an undefined or empty value because it could be interpreted as
+             * a 'valid' key hash on the reader side
+             **/
             fastdds::dds::ParameterSerializer<fastdds::dds::Parameter_t>::add_parameter_key(msg,
                     change->instanceHandle);
 
+            /** Changes like UNREGISTER or DISPOSE must include the key, they can't be sent without key */
```

Comment:
We should also check whether this is true.
Might be the case that we can send the status parameter without the hash one if the handle is not set but the payload is present.